Source/FieldSolver/MagnetostaticSolver/MagnetostaticSolver.cpp

```diff
@@ -0,0 +1,462 @@
+                                                                       const std::unique_ptr<amrex::MultiFab> &dst)
+{
+    WarpX &warpx = WarpX::GetInstance();
+
+    // Grab Interpolation Coefficients
+    // Order of finite-order centering of fields
+    const int fg_nox = WarpX::field_centering_nox;
+    const int fg_noy = WarpX::field_centering_noy;
+    const int fg_noz = WarpX::field_centering_noz;
+
+    // Device vectors of stencil coefficients used for finite-order centering of fields
+    amrex::Real const * stencil_coeffs_x = warpx.device_field_centering_stencil_coeffs_x.data();
+    amrex::Real const * stencil_coeffs_y = warpx.device_field_centering_stencil_coeffs_y.data();
+    amrex::Real const * stencil_coeffs_z = warpx.device_field_centering_stencil_coeffs_z.data();
+
+    // Synchronize the ghost cells, do halo exchange
+    ablastr::utils::communication::FillBoundary(*src,
+                                                src->nGrowVect(),
+                                                WarpX::do_single_precision_comms);
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+    for (MFIter mfi(*dst, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+    {
+        IntVect const src_stag = src->ixType().toIntVect();
+        IntVect const dst_stag = dst->ixType().toIntVect();
+
+        Array4<amrex::Real const> const& src_arr = src->const_array(mfi);
+        Array4<amrex::Real> const& dst_arr = dst->array(mfi);
+
+        Box bx = mfi.growntilebox(dst->nGrowVect());
+
+        ParallelFor(bx, [=] AMREX_GPU_DEVICE (int j, int k, int l) noexcept
+        {
+            warpx_interp(j, k, l, dst_arr, src_arr, dst_stag, src_stag, fg_nox, fg_noy, fg_noz,
+                stencil_coeffs_x, stencil_coeffs_y, stencil_coeffs_z);
+        });
+    }
+
+    // Synchronize the ghost cells, do halo exchange
+    ablastr::utils::communication::FillBoundary(*dst,
+                                                dst->nGrowVect(),
+                                                WarpX::do_single_precision_comms);
+}
+
+void MagnetostaticSolver::EBCalcBfromVectorPotentialPerLevel::operator()(amrex::Array<std::unique_ptr<amrex::MLMG>,3> & mlmg, int const lev)
+{
+    using namespace amrex::literals;
+
+    // This operator gets the gradient solution on the cell edges, aligned with E field staggered grid
+    // This routine interpolates to the B-field staggered grid,
+
+    amrex::Array<amrex::MultiFab*, AMREX_SPACEDIM> buf_ptr =
+    {
+#if defined(WARPX_DIM_3D)
+        m_grad_buf_e_stag[lev][0].get(),
+        m_grad_buf_e_stag[lev][1].get(),
+        m_grad_buf_e_stag[lev][2].get()
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+        m_grad_buf_e_stag[lev][0].get(),
+        m_grad_buf_e_stag[lev][2].get()
+#endif
+    };
+
+    // This will grab the gradient values for Ax
+    mlmg[0]->getGradSolution({buf_ptr});
```

Comment:
This function call needs to fill the guard cells (at least one cell deep) in order for the following `doInterp` function to work properly ; do you know whether that is the case?

Reply:
I believe I was able to fix this by doing a halo exchange in the doInterp function.

Reply:
Ah, yes, sorry I overlooked this.